File: tests/oracle/rule_collect_terms_1204.cc
/*
 * Locked Ghidra 12.0.4 RuleCollectTerms::applyOp oracle.
 *
 * Every case records the target-relevant structural IR immediately before and
 * after the real rule call: block and op-bank alive/dead order,
 * op lifecycle/parent/order, every
 * input and output, and the def-use state of both still-reachable and detached
 * Varnodes. Unique-space offsets are deliberately represented by stable
 * fixture identities because they are allocation-only temporaries. The narrow
 * fixture does not claim to observe unrelated type/symbol/high state.
 */

#include "bfd_arch.hh"
#include "libdecomp.hh"
#include "ruleaction.hh"

#include <iomanip>
#include <iostream>
#include <iterator>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::map;
using std::ostringstream;
using std::string;
using std::vector;

class Fixture {
  Funcdata &fd;
  BlockBasic *block;
  vector<PcodeOp *> ops;
  map<PcodeOp *,string> opNames;
  vector<Varnode *> varnodes;
  map<Varnode *,string> varnodeNames;
  int4 nextOp;
  int4 nextVarnode;

  void rememberOp(PcodeOp *op,const string &name)
  {
    if (opNames.insert(std::make_pair(op,name)).second)
      ops.push_back(op);
  }

  void rememberVarnode(Varnode *vn,const string &name)
  {
    if (varnodeNames.insert(std::make_pair(vn,name)).second)
      varnodes.push_back(vn);
  }

  string opName(PcodeOp *op) const
  {
    map<PcodeOp *,string>::const_iterator iter = opNames.find(op);
    if (iter == opNames.end())
      throw std::runtime_error("unregistered fixture op");
    return (*iter).second;
  }

  string varnodeName(Varnode *vn) const
  {
    map<Varnode *,string>::const_iterator iter = varnodeNames.find(vn);
    if (iter == varnodeNames.end())
      throw std::runtime_error("unregistered fixture varnode");
    return (*iter).second;
  }

  void discover(void)
  {
    for(list<PcodeOp *>::const_iterator iter=block->beginOp();
        iter!=block->endOp();++iter) {
      if (opNames.find(*iter) == opNames.end()) {
        ostringstream name;
        name << 'n' << nextOp++;
        rememberOp(*iter,name.str());
      }
    }
    for(list<PcodeOp *>::const_iterator iter=block->beginOp();
        iter!=block->endOp();++iter) {
      PcodeOp *op = *iter;
      if (op->getOut() != (Varnode *)0 &&
          varnodeNames.find(op->getOut()) == varnodeNames.end())
        rememberVarnode(op->getOut(),opName(op) + "_out");
      for(int4 slot=0;slot<op->numInput();++slot) {
        Varnode *vn = op->getIn(slot);
        if (varnodeNames.find(vn) == varnodeNames.end()) {
          ostringstream name;
          name << 'g' << nextVarnode++;
          rememberVarnode(vn,name.str());
        }
      }
    }
  }

  string opList(list<PcodeOp *>::const_iterator iter,
                list<PcodeOp *>::const_iterator enditer) const
  {
    ostringstream out;
    bool first = true;
    for(;iter!=enditer;++iter) {
      if (!first) out << ',';
      first = false;
      out << opName(*iter);
    }
    return out.str();
  }

  string blockOrder(void) const
  {
    return opList(block->beginOp(),block->endOp());
  }

  string opState(PcodeOp *op) const
  {
    ostringstream out;
    out << opName(op)
        << "{opc=" << static_cast<int4>(op->code())
        << ",dead=" << op->isDead()
        << ",parent=" << (op->getParent() == block ? block->getIndex() : -1)
        << ",order=" << op->getSeqNum().getOrder()
        << ",inputs=[";
    for(int4 slot=0;slot<op->numInput();++slot) {
      if (slot != 0) out << ',';
      out << varnodeName(op->getIn(slot));
    }
    out << "],output=";
    if (op->getOut() == (Varnode *)0)
      out << '-';
    else
      out << varnodeName(op->getOut());
    out << '}';
    return out.str();
  }

  string varnodeState(Varnode *vn) const
  {
    ostringstream out;
    out << varnodeName(vn)
        << "{space=" << vn->getSpace()->getName()
        << ",size=" << vn->getSize()
        << ",offset=";
    if (vn->getSpace()->getType() == IPTR_INTERNAL)
      out << "tmp";
    else
      out << std::hex << vn->getOffset() << std::dec;
    out << ",constant=" << vn->isConstant()
        << ",input=" << vn->isInput()
        << ",written=" << vn->isWritten()
        << ",free=" << vn->isFree()
        << ",def=";
    if (vn->getDef() == (PcodeOp *)0)
      out << '-';
    else
      out << opName(vn->getDef());
    out << ",desc=[";
    bool first = true;
    for(list<PcodeOp *>::const_iterator iter=vn->beginDescend();
        iter!=vn->endDescend();++iter) {
      if (!first) out << ',';
      first = false;
      out << opName(*iter) << '.' << (*iter)->getSlot(vn);
    }
    out << "]}";
    return out.str();
  }

public:
  explicit Fixture(Funcdata &func)
    : fd(func), block((BlockBasic *)0), nextOp(0), nextVarnode(0)
  {
    BlockGraph &graph = const_cast<BlockGraph &>(fd.getBasicBlocks());
    block = graph.newBlockBasic(&fd);
  }

  Varnode *makeInput(const string &name,int4 size,uintb offset)
  {
    AddrSpace *registerSpace = fd.getArch()->getSpaceByName("register");
    if (registerSpace == (AddrSpace *)0)
      throw std::runtime_error("fixture requires the register space");
    Varnode *vn = fd.setInputVarnode(fd.newVarnode(size,registerSpace,offset));
    rememberVarnode(vn,name);
    return vn;
  }

  Varnode *makeConstant(const string &name,int4 size,uintb value)
  {
    Varnode *vn = fd.newConstant(size,value);
    rememberVarnode(vn,name);
    return vn;
  }

  PcodeOp *makeOp(const string &name,OpCode opcode,int4 inputs,int4 outputSize)
  {
    AddrSpace *codeSpace = fd.getArch()->getDefaultCodeSpace();
    PcodeOp *op = fd.newOp(inputs,Address(codeSpace,0x1000));
    fd.opSetOpcode(op,opcode);
    rememberOp(op,name);
    Varnode *outvn = fd.newUniqueOut(outputSize,op);
    rememberVarnode(outvn,name + "_out");
    return op;
  }

  void setInput(PcodeOp *op,Varnode *vn,int4 slot)
  {
    fd.opSetInput(op,vn,slot);
  }

  void insertEnd(PcodeOp *op)
  {
    fd.opInsertEnd(op,block);
  }

  void dump(const string &caseName,const string &stage,const string &result)
  {
    discover();
    ostringstream opStates;
    bool first = true;
    for(list<PcodeOp *>::const_iterator iter=block->beginOp();
        iter!=block->endOp();++iter) {
      if (!first) opStates << ';';
      first = false;
      opStates << opState(*iter);
    }

    ostringstream varnodeStates;
    first = true;
    for(vector<Varnode *>::const_iterator iter=varnodes.begin();
        iter!=varnodes.end();++iter) {
      if (!first) varnodeStates << ';';
      first = false;
      varnodeStates << varnodeState(*iter);
    }

    std::cout << "case=" << caseName
              << "|stage=" << stage
              << "|result=" << result
              << "|block=[" << blockOrder() << ']'
              << "|alive=[" << opList(fd.beginOpAlive(),fd.endOpAlive()) << ']'
              << "|dead=[" << opList(fd.beginOpDead(),fd.endOpDead()) << ']'
              << "|ops=[" << opStates.str() << ']'
              << "|varnodes=[" << varnodeStates.str() << "]\n";
  }
};

void runLikeCase(Funcdata &fd,const string &name,int4 size,uintb coef0,uintb coef1)
{
  fd.clear();
  Fixture fixture(fd);
  Varnode *x = fixture.makeInput("x",size,0x40);
  PcodeOp *mult0 = fixture.makeOp("mult0",CPUI_INT_MULT,2,size);
  PcodeOp *mult1 = fixture.makeOp("mult1",CPUI_INT_MULT,2,size);
  PcodeOp *root = fixture.makeOp("root",CPUI_INT_ADD,2,size);
  Varnode *constant0 = fixture.makeConstant("coef0",size,coef0);
  Varnode *constant1 = fixture.makeConstant("coef1",size,coef1);

  fixture.setInput(mult0,x,0);
  fixture.setInput(mult0,constant0,1);
  fixture.setInput(mult1,x,0);
  fixture.setInput(mult1,constant1,1);
  fixture.setInput(root,mult0->getOut(),0);
  fixture.setInput(root,mult1->getOut(),1);
  fixture.insertEnd(mult0);
  fixture.insertEnd(mult1);
  fixture.insertEnd(root);

  fixture.dump(name,"before","na");
  RuleCollectTerms rule("analysis");
  int4 result = rule.applyOp(root,fd);
  fixture.dump(name,"after",std::to_string(result));
}

void runConstantCase(Funcdata &fd,const string &name,int4 size,uintb innerValue,
                     uintb rootValue)
{
  fd.clear();
  Fixture fixture(fd);
  Varnode *x = fixture.makeInput("x",size,0x40);
  // Keep creation and fixture-identity registration order identical on both
  // sides. Ghidra's termOrder deliberately compares any two constants equal.
  Varnode *rootConstant = fixture.makeConstant("root_const",size,rootValue);
  Varnode *innerConstant = fixture.makeConstant("inner_const",size,innerValue);
  PcodeOp *inner = fixture.makeOp("inner",CPUI_INT_ADD,2,size);
  PcodeOp *root = fixture.makeOp("root",CPUI_INT_ADD,2,size);

  fixture.setInput(inner,x,0);
  fixture.setInput(inner,innerConstant,1);
  fixture.setInput(root,inner->getOut(),0);
  fixture.setInput(root,rootConstant,1);
  fixture.insertEnd(inner);
  fixture.insertEnd(root);

  fixture.dump(name,"before","na");
  RuleCollectTerms rule("analysis");
  int4 result = rule.applyOp(root,fd);
  fixture.dump(name,"after",std::to_string(result));
}

void run(const string &specDirectory,const string &binary)
{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  {
    BfdArchitecture architecture(binary,"default",&std::cerr);
    DocumentStorage store;
    architecture.init(store);
    architecture.readLoaderSymbols("::");
    Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
    if (fd == (Funcdata *)0)
      throw std::runtime_error("GetStr was not found in the BFD symbol table");
    if (fd->getName() != "GetStr" ||
        fd->getAddress().getOffset() != 0x36d0 || fd->getSize() != 0)
      throw std::runtime_error("GetStr input identity drifted");
    if (architecture.archid != "x86:LE:64:default:gcc")
      throw std::runtime_error("runtime architecture/compiler drifted: " +
                               architecture.archid);

    runLikeCase(*fd,"like_nonoverflow",8,2,3);
    runLikeCase(*fd,"like_storage_wrap",1,0xff,2);
    runLikeCase(*fd,"like_uintb_wrap",8,~((uintb)0),2);
    runLikeCase(*fd,"like_zero",1,0xff,1);
    runConstantCase(*fd,"constant_nonoverflow",8,3,5);
    runConstantCase(*fd,"constant_storage_wrap",1,0xff,2);
    runConstantCase(*fd,"constant_uintb_wrap",8,~((uintb)0),2);
  }
  shutdownDecompilerLibrary();
}

} // anonymous namespace

int main(int argc,char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: rule_collect_terms_1204 SPEC_ROOT CURL_BINARY\n";
    return 2;
  }
  try {
    run(argv[1],argv[2]);
    return 0;
  }
  catch(const ghidra::LowlevelError &error) {
    std::cerr << "Ghidra LowlevelError: " << error.explain << '\n';
  }
  catch(const ghidra::DecoderError &error) {
    std::cerr << "Ghidra DecoderError: " << error.explain << '\n';
  }
  catch(const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/rule_collect_terms_1204.rs
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

use rugra::action::Rule;
use rugra::address::Address;
use rugra::funcdata::Funcdata;
use rugra::op::{PcodeOp, PcodeOpRef};
use rugra::opcodes::OpCode;
use rugra::ruleaction::RuleCollectTerms;
use rugra::space::AddressSpace;
use rugra::varnode::Varnode;

type VarnodeRef = Arc<RwLock<Varnode>>;

struct Fixture {
    fd: Funcdata,
    block: Arc<RwLock<dyn rugra::block::FlowBlock + Send + Sync>>,
    ops: Vec<PcodeOpRef>,
    op_names: HashMap<usize, String>,
    varnodes: Vec<VarnodeRef>,
    varnode_names: HashMap<usize, String>,
    next_op: usize,
    next_varnode: usize,
}

impl Fixture {
    fn new() -> Self {
        let mut fd = Funcdata::new("GetStr", Address::new(0x36d0), 0);
        assert_eq!(fd.name, "GetStr");
        assert_eq!(fd.baseaddr.as_u64(), 0x36d0);
        assert_eq!(fd.size, 0);
        let block = fd.create_new_block();
        Self {
            fd,
            block,
            ops: Vec::new(),
            op_names: HashMap::new(),
            varnodes: Vec::new(),
            varnode_names: HashMap::new(),
            next_op: 0,
            next_varnode: 0,
        }
    }

    fn op_key(op: &PcodeOpRef) -> usize {
        Arc::as_ptr(&op.0) as usize
    }

    fn op_arc_key(op: &Arc<RwLock<PcodeOp>>) -> usize {
        Arc::as_ptr(op) as usize
    }

    fn varnode_key(vn: &VarnodeRef) -> usize {
        Arc::as_ptr(vn) as usize
    }

    fn remember_op(&mut self, op: PcodeOpRef, name: String) {
        let key = Self::op_key(&op);
        if self.op_names.insert(key, name).is_none() {
            self.ops.push(op);
        }
    }

    fn remember_varnode(&mut self, vn: VarnodeRef, name: String) {
        let key = Self::varnode_key(&vn);
        if self.varnode_names.insert(key, name).is_none() {
            self.varnodes.push(vn);
        }
    }

    fn op_name(&self, op: &PcodeOpRef) -> &str {
        self.op_names
            .get(&Self::op_key(op))
            .expect("registered fixture op")
    }

    fn op_arc_name(&self, op: &Arc<RwLock<PcodeOp>>) -> &str {
        self.op_names
            .get(&Self::op_arc_key(op))
            .expect("registered fixture op")
    }

    fn varnode_name(&self, vn: &VarnodeRef) -> &str {
        self.varnode_names
            .get(&Self::varnode_key(vn))
            .expect("registered fixture varnode")
    }

    fn discover(&mut self) {
        let block_ops = self.block.read().unwrap().get_ops();
        for op in &block_ops {
            if !self.op_names.contains_key(&Self::op_key(op)) {
                let name = format!("n{}", self.next_op);
                self.next_op += 1;
                self.remember_op(op.clone(), name);
            }
        }
        for op in &block_ops {
            let op_name = self.op_name(op).to_string();
            let (output, inputs) = {
                let op = op.0.read().unwrap();
                (op.output.clone(), op.inrefs.clone())
            };
            if let Some(output) = output {
                if !self.varnode_names.contains_key(&Self::varnode_key(&output)) {
                    self.remember_varnode(output, format!("{op_name}_out"));
                }
            }
            for input in inputs {
                if !self.varnode_names.contains_key(&Self::varnode_key(&input)) {
                    let name = format!("g{}", self.next_varnode);
                    self.next_varnode += 1;
                    self.remember_varnode(input, name);
                }
            }
        }
    }

    fn make_input(&mut self, name: &str, size: usize, offset: u64) -> VarnodeRef {
        let vn = self
            .fd
            .vbank
            .create_with_space(size, AddressSpace::Register, offset);
        let vn = self.fd.set_input_varnode(vn);
        self.remember_varnode(vn.clone(), name.to_string());
        vn
    }

    fn make_constant(&mut self, name: &str, size: usize, value: u64) -> VarnodeRef {
        let vn = self.fd.new_constant(size, value);
        self.remember_varnode(vn.clone(), name.to_string());
        vn
    }

    fn make_op(
        &mut self,
        name: &str,
        opcode: OpCode,
        inputs: usize,
        output_size: usize,
    ) -> PcodeOpRef {
        let op = self.fd.new_op(inputs, Address::new(0x1000));
        self.fd.op_set_opcode(&op, opcode);
        self.remember_op(op.clone(), name.to_string());
        let output = self.fd.new_unique_out(output_size, &op);
        self.remember_varnode(output, format!("{name}_out"));
        op
    }

    fn set_input(&mut self, op: &PcodeOpRef, vn: VarnodeRef, slot: usize) {
        self.fd.op_set_input(op, vn, slot);
    }

    fn insert_end(&mut self, op: &PcodeOpRef) {
        self.fd.op_insert_end(op, &self.block.clone());
    }

    fn op_list(&self, ops: &[PcodeOpRef]) -> String {
        ops.iter()
            .map(|op| self.op_name(op).to_string())
            .collect::<Vec<_>>()
            .join(",")
    }

    fn op_state(&self, op: &PcodeOpRef) -> String {
        let op_guard = op.0.read().unwrap();
        let parent = op_guard
            .parent
            .as_ref()
            .and_then(std::sync::Weak::upgrade)
            .map(|parent| parent.read().unwrap().get_index())
            .unwrap_or(-1);
        let inputs = op_guard
            .inrefs
            .iter()
            .map(|vn| self.varnode_name(vn).to_string())
            .collect::<Vec<_>>()
            .join(",");
        let output = op_guard
            .output
            .as_ref()
            .map(|vn| self.varnode_name(vn).to_string())
            .unwrap_or_else(|| "-".to_string());
        format!(
            "{}{{opc={},dead={},parent={},order={},inputs=[{}],output={}}}",
            self.op_name(op),
            op_guard.opcode as i32,
            u8::from(op_guard.is_dead()),
            parent,
            op_guard.start.get_order(),
            inputs,
            output,
        )
    }

    fn varnode_state(&self, vn: &VarnodeRef) -> String {
        let vn_guard = vn.read().unwrap();
        let space = vn_guard.get_space();
        let offset = if space == AddressSpace::Unique {
            "tmp".to_string()
        } else {
            format!("{:x}", vn_guard.get_offset())
        };
        let definition = vn_guard
            .def
            .as_ref()
            .and_then(std::sync::Weak::upgrade)
            .map(|op| self.op_arc_name(&op).to_string())
            .unwrap_or_else(|| "-".to_string());
        let descendants = vn_guard
            .descend
            .iter()
            .filter_map(std::sync::Weak::upgrade)
            .map(|op| {
                let slot = op
                    .read()
                    .unwrap()
                    .inrefs
                    .iter()
                    .position(|input| Arc::ptr_eq(input, vn))
                    .expect("descendant retains input varnode");
                format!("{}.{}", self.op_arc_name(&op), slot)
            })
            .collect::<Vec<_>>()
            .join(",");
        format!(
            "{}{{space={},size={},offset={},constant={},input={},written={},free={},def={},desc=[{}]}}",
            self.varnode_name(vn),
            space.name(),
            vn_guard.get_size(),
            offset,
            u8::from(vn_guard.is_constant()),
            u8::from(vn_guard.is_input()),
            u8::from(vn_guard.is_written()),
            u8::from(vn_guard.is_free()),
            definition,
            descendants,
        )
    }

    fn dump(&mut self, case_name: &str, stage: &str, result: &str) {
        self.discover();
        let block_ops = self.block.read().unwrap().get_ops();
        let op_states = block_ops
            .iter()
            .map(|op| self.op_state(op))
            .collect::<Vec<_>>()
            .join(";");
        let varnode_states = self
            .varnodes
            .iter()
            .map(|vn| self.varnode_state(vn))
            .collect::<Vec<_>>()
            .join(";");
        println!(
            "case={case_name}|stage={stage}|result={result}|block=[{}]|alive=[{}]|dead=[{}]|ops=[{}]|varnodes=[{}]",
            self.op_list(&block_ops),
            self.op_list(&self.fd.obank.alivelist),
            self.op_list(&self.fd.obank.deadlist),
            op_states,
            varnode_states,
        );
    }
}

fn run_like_case(name: &str, size: usize, coef0: u64, coef1: u64) {
    let mut fixture = Fixture::new();
    let x = fixture.make_input("x", size, 0x40);
    let mult0 = fixture.make_op("mult0", OpCode::CPUI_INT_MULT, 2, size);
    let mult1 = fixture.make_op("mult1", OpCode::CPUI_INT_MULT, 2, size);
    let root = fixture.make_op("root", OpCode::CPUI_INT_ADD, 2, size);
    let constant0 = fixture.make_constant("coef0", size, coef0);
    let constant1 = fixture.make_constant("coef1", size, coef1);

    fixture.set_input(&mult0, x.clone(), 0);
    fixture.set_input(&mult0, constant0, 1);
    fixture.set_input(&mult1, x, 0);
    fixture.set_input(&mult1, constant1, 1);
    let mult0_out = mult0.0.read().unwrap().output.clone().unwrap();
    let mult1_out = mult1.0.read().unwrap().output.clone().unwrap();
    fixture.set_input(&root, mult0_out, 0);
    fixture.set_input(&root, mult1_out, 1);
    fixture.insert_end(&mult0);
    fixture.insert_end(&mult1);
    fixture.insert_end(&root);

    fixture.dump(name, "before", "na");
    let result = RuleCollectTerms::new()
        .apply_op(&root.0, &mut fixture.fd)
        .expect("RuleCollectTerms like-term case");
    fixture.dump(name, "after", &result.to_string());
}

fn run_constant_case(name: &str, size: usize, inner_value: u64, root_value: u64) {
    let mut fixture = Fixture::new();
    let x = fixture.make_input("x", size, 0x40);
    // Keep creation and fixture-identity registration order identical on both
    // sides. Ghidra's termOrder deliberately compares any two constants equal.
    let root_constant = fixture.make_constant("root_const", size, root_value);
    let inner_constant = fixture.make_constant("inner_const", size, inner_value);
    let inner = fixture.make_op("inner", OpCode::CPUI_INT_ADD, 2, size);
    let root = fixture.make_op("root", OpCode::CPUI_INT_ADD, 2, size);

    fixture.set_input(&inner, x, 0);
    fixture.set_input(&inner, inner_constant, 1);
    let inner_out = inner.0.read().unwrap().output.clone().unwrap();
    fixture.set_input(&root, inner_out, 0);
    fixture.set_input(&root, root_constant, 1);
    fixture.insert_end(&inner);
    fixture.insert_end(&root);

    fixture.dump(name, "before", "na");
    let result = RuleCollectTerms::new()
        .apply_op(&root.0, &mut fixture.fd)
        .expect("RuleCollectTerms constant case");
    fixture.dump(name, "after", &result.to_string());
}

fn main() {
    run_like_case("like_nonoverflow", 8, 2, 3);
    run_like_case("like_storage_wrap", 1, 0xff, 2);
    run_like_case("like_uintb_wrap", 8, u64::MAX, 2);
    run_like_case("like_zero", 1, 0xff, 1);
    run_constant_case("constant_nonoverflow", 8, 3, 5);
    run_constant_case("constant_storage_wrap", 1, 0xff, 2);
    run_constant_case("constant_uintb_wrap", 8, u64::MAX, 2);
}
